crates/core/src/bc_protocol/connection/udpconn/discover.rs

```diff
@@ -0,0 +1,543 @@
+        // let local_port = local_addr.port();
+
+        let mtu = MTU;
+
+        for p2p_relay in P2P_RELAY_HOSTNAMES.iter() {
+            debug!("Trying register: {}", p2p_relay);
+            let msg = BcUdp::Discovery(UdpDiscovery {
+                tid,
+                payload: UdpXml {
+                    c2m_q: Some(C2mQ {
+                        uid: uid.to_string(),
+                        os: "MAC".to_string(),
+                    }),
+                    ..Default::default()
+                },
+            });
+
+            let mut buf = vec![];
+            msg.serialize(&mut buf)?;
+
+            let (amount_sent, abort) =
+                Self::retrying_send_to(socket, &buf[..], format!("{}:9999", p2p_relay))?;
+            assert_eq!(amount_sent, buf.len());
+
+            let start_time = OffsetDateTime::now_utc();
+            loop {
+                if (OffsetDateTime::now_utc() - start_time) <= timeout {
+                    let mut buf = vec![0; mtu as usize];
+                    if let Ok((bytes_read, _)) = socket.recv_from(&mut buf) {
+                        match BcUdp::deserialize(&buf[0..bytes_read]) {
+                            Ok(BcUdp::Discovery(UdpDiscovery {
+                                tid: _,
+                                payload:
+                                    UdpXml {
+                                        m2c_q_r: Some(m2c_q_r),
+                                        ..
+                                    },
+                            })) if m2c_q_r.reg.port == 0 || m2c_q_r.reg.ip.is_empty() => {
+                                // This register is empty
+                                abort.abort();
+                                break;
+                            }
+                            Ok(BcUdp::Discovery(UdpDiscovery {
+                                tid: _,
+                                payload:
+                                    UdpXml {
+                                        m2c_q_r: Some(m2c_q_r),
+                                        ..
+                                    },
+                            })) => {
+                                abort.abort();
+                                return Ok(m2c_q_r);
+                            }
+                            Ok(smt) => debug!("Udp Discovery got this unexpected BcUdp {:?}", smt),
+                            Err(_) => debug!(
+                                "Udp Discovery got this unexpected binary: {:?}",
+                                &buf[0..bytes_read]
+                            ),
+                        }
+                    }
+                } else {
+                    abort.abort();
+                    // This p2p server didn't have the data we need try another
+                    break;
+                }
+            }
+        }
+        debug!("All registers tried");
+        Err(Error::ConnectionUnavaliable)
+    }
+
+    fn discover_from_uuid_remote(
```

Comment:
```suggestion
    // This will start remote discovery against the reolink p2p servers
    //
    // It has the following stages
    // - Contact the p2p register: This will get the register, relay and log server ip addresses
    // - Register the client ip address to the register (This can probably be used for NAT holepunching if the camera
    //    tries to connect to us using this data, currently we don't register a real address though, just `0.0.0.0`)
    // - The register will return the camera ip/port and SID
    // - Connect to the camera using register's provided ip/port and negotiate a cid and did
    // - Send a message to the reolink log server that we will connect locally
    // - Send a message to the camera that we could connect remotely
    fn discover_from_uuid_remote(
```